skt/misc.py

```diff
@@ -24,6 +25,44 @@
 LOGGER = logging.getLogger()
 
 
+@retry(RuntimeError)
+def retry_safe_popen(err_exc_strings, *args, stdin_data=None, **kwargs):
+    """ Call safe_popen with *args, stdin_data=None, **kwargs provided.
+        If stderr stream is present and contains any string in err_exc_strings
+        list, then the process call is done again with retry after 3 seconds
+        (see retry decorator). Log commands retried and allow 3 retries max.
+        Also log if last command failed and we gave up. The program execution
+        is not terminated / no exception is raised on last failure.
+
+        Args:
+            err_exc_strings: a list of strings; if any is present in stderr,
+                             retry the command
+            args:            arguments to pass to Popen
+            stdin_data:      None or str, use None when you don't want to pass
+                             string data to stdin
+            kwargs:          keyword arguments to pass to Popen
+        Returns:
+            tuple (stdout, stderr, returncode) where
+                stdout is a string
+                stderr is a string
+                returncode is an integer
+    """
+    stdout, stderr, returncode = safe_popen(*args, stdin_data=stdin_data,
+                                            **kwargs)
+    for err_str in err_exc_strings:
+        if stderr and err_str in stderr:
+            logging.warning(stderr)
+            safe_popen_args = ' '.join(list(*args))
+            logging.warning('RETRY shell command "%s"', safe_popen_args)
+            if not (retry_safe_popen.failed_count + 1 ==
```

Comment:
Isn't this a bit clearer as `if retry_safe_popen.failed_count + 1 < retry_safe_popen.retries`?

And maybe add a comment stating that `.failed_count` doesn't count the current error, hence the need for the `+ 1`. It took me five minutes of back and forth to figure it out, I almost write that there was a bug in the code 😅